---
title: "quant_gen_project"
author: "Darya Akimova"
date: "April 20, 2018"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r packages, comment=NA, include=FALSE}
library(dplyr)  # for data manipulation
library(tidyr)  # for data manipulation
library(tibble)  # for data manipulation
library(broom)  # for cleaning up models
library(purrr)  # for applying functions
library(ggplot2)  # for plotting
library(cowplot)  # for plotting, modifies ggplot2 defaults
library(GGally)  # for plotting many variables together
library(MASS)  # for access to matrix functions
```

```{r data, comment=NA, include=FALSE}
# read in data
# geno = all provided genotype data
geno <- read.csv("QG18_genotypes.csv", row.names = 1)
# pheno = all provided phenotype data
pheno <- read.csv("QG18_phenotypes.csv", row.names = 1)
# covars = all information on covariates (Population/genetic background and Sex of individuals)
covars <- read.csv("QG18_covars.csv", row.names = 1, stringsAsFactors = FALSE)
# snp_info = information on all snps (chromosome and position)
snp_info <- read.csv("QG18_SNP_info.csv", stringsAsFactors = FALSE)
# gene_info = information on genes that the expression levels were measured as the phenotype
gene_info <- read.csv("QG18_gene_info.csv", stringsAsFactors = FALSE)
```

```{r data_quality_checks, comment=NA}
# is there any missing data?
anyNA(list(geno, pheno, covars, snp_info, gene_info))
# are there approximately equal numbers of people in each covariate group?
knitr::kable(table(covars$Population))
knitr::kable(table(covars$Sex))
# does the coding of the genotypes makes sense across all the data?
max(as.matrix(geno))
min(as.matrix(geno))
# calculate allele frequency
geno_sums <- map_dbl(geno, function(x) sum(x) / (nrow(geno) * 2))
# any genotypes need to be removed because of MAF < 5%?
any(geno_sums < 0.05 | geno_sums > 0.95)
# no genotypes need to be removed because of MAF < 5%
```

Phenotype plots:
```{r comment=NA}
pheno %>% 
  rownames_to_column("sample") %>% 
  gather("probe", "abundance", 2:6) %>% 
  left_join(gene_info, by = "probe") %>% 
  ggplot(aes(x = abundance, fill = symbol)) +
  geom_histogram(binwidth = 0.1) +
  facet_wrap(~ symbol, nrow = 1) 
pheno %>% 
  rownames_to_column("sample") %>% 
  gather("probe", "abundance", 2:6) %>% 
  left_join(gene_info, by = "probe") %>% 
  left_join(
    covars %>% rownames_to_column("sample"), 
    by = "sample") %>% 
  ggplot(aes(x = abundance, fill = Population)) +
  geom_histogram(binwidth = 0.1, alpha = 0.8, position = "identity") +
  facet_grid(Population ~ symbol)
pheno %>% 
  rownames_to_column("sample") %>% 
  gather("probe", "abundance", 2:6) %>% 
  left_join(gene_info, by = "probe") %>% 
  left_join(
    covars %>% rownames_to_column("sample"), 
    by = "sample") %>% 
  ggplot(aes(x = abundance, fill = Sex)) +
  geom_histogram(binwidth = 0.1, alpha = 0.8, position = "identity") +
  facet_grid(Sex ~ symbol)
pheno %>% 
  rownames_to_column("sample") %>% 
  gather("probe", "abundance", 2:6) %>% 
  left_join(gene_info, by = "probe") %>% 
  left_join(
    covars %>% rownames_to_column("sample"), 
    by = "sample") %>% 
  ggplot(aes(x = Sex, y = abundance, fill = Sex)) +
  geom_boxplot() +
  facet_wrap(~ symbol)
pheno %>% 
  rownames_to_column("sample") %>% 
  gather("probe", "abundance", 2:6) %>% 
  left_join(gene_info, by = "probe") %>% 
  left_join(
    covars %>% rownames_to_column("sample"), 
    by = "sample") %>% 
  ggplot(aes(x = Population, y = abundance, fill = Population)) +
  geom_boxplot() +
  facet_wrap(~ symbol)

# are there any genotypes without associated phenotypes, or vice versa?
all.equal(rownames(geno), rownames(pheno))
pheno_pca <- prcomp(pheno)
pheno_pca_x <- as.data.frame(pheno_pca$x)
pheno_pca_x %>% 
  rownames_to_column("sample") %>% 
  left_join(covars %>% rownames_to_column("sample"), by = "sample") %>% 
  ggplot(aes(x = PC1, y = PC2, color = Population)) +
  geom_point(size = 3, alpha = 0.5)
ggpairs(pheno)
```

Genotype plots:
```{r comment=NA}
# PCA analysis of genetic data
geno_pca <- prcomp(geno, center = TRUE, scale. = TRUE)
plot((geno_pca$sdev^2/ sum(geno_pca$sdev^2)) * 100, xlab = "Principal Component", ylab = "Percent of total variance explained")
geno_pca_x <- data.frame(geno_pca$x)
geno_pca_x %>% 
  rownames_to_column("sample") %>% 
  left_join(covars %>% rownames_to_column("sample"), by = "sample") %>% 
  ggplot(aes(x = PC1, y = PC2, color = Sex)) +
    geom_point(size = 2, alpha = 0.8) +
    ggtitle("First two principal components, colored by Sex")
geno_pca_x %>% 
  rownames_to_column("sample") %>% 
  left_join(covars %>% rownames_to_column("sample"), by = "sample") %>% 
  ggplot(aes(x = PC1, y = PC2, color = Population)) +
    geom_point(size = 2, alpha = 0.5) +
    ggtitle("First two principal components, colored by Population") +
  ylim(-70, 30)
#indv_pca <- prcomp(t(geno), center = TRUE, scale. = TRUE)
#indv_pca_x <- indv_pca$x
```

Based on on the principal component analysis of the genomes, colored by the population of origin, it is possible to see that there is clearly population structure. 

Test each covariate individually. 

```{r comment=NA}
x_a <- as.matrix(geno - 1)
x_d <- replace(as.matrix(geno), which(as.matrix(geno) == 2 | as.matrix(geno) == 0), -1)
MLE <- function(xa, xd, y_mat) {
  X_mat <- cbind(1, xa, xd)
  beta_hat <- ginv(t(X_mat) %*% X_mat) %*% t(X_mat) %*% y_mat
}
fstat_calc <- function(xa, xd, MLE, y_mat) {
  X_mat <- cbind(1, xa, xd)
  y_hat <- X_mat %*% MLE
  SSM <- sum((y_hat - mean(y_mat)) ^ 2)
  SSE <- sum((y_mat - y_hat) ^ 2)
  df_M <- 2
  df_E <- length(xa) - 3
  Fstat <- (SSM / df_M) / (SSE / df_E)
  return(Fstat)
}

# gene 1
mle_gene1 <- map2(data.frame(x_a), data.frame(x_d), MLE, as.numeric(as.matrix(pheno[, 1])))
pval_gene1 <- pmap(
  list(
    xa = data.frame(x_a), 
    xd = data.frame(x_d), 
    MLE = mle_gene1
    ), 
  fstat_calc, pheno[, 1]
  ) %>% 
  flatten_dbl() %>% 
  pf(2, 344 - 3, lower.tail = FALSE)
# gene 2
mle_gene2 <- map2(data.frame(x_a), data.frame(x_d), MLE, as.numeric(as.matrix(pheno[, 2])))
pval_gene2 <- pmap(
  list(
    xa = data.frame(x_a), 
    xd = data.frame(x_d), 
    MLE = mle_gene2
    ), 
  fstat_calc, pheno[, 2]
  ) %>% 
  flatten_dbl() %>% 
  pf(2, 344 - 3, lower.tail = FALSE)
# gene 3
mle_gene3 <- map2(data.frame(x_a), data.frame(x_d), MLE, as.numeric(as.matrix(pheno[, 3])))
pval_gene3 <- pmap(
  list(
    xa = data.frame(x_a), 
    xd = data.frame(x_d), 
    MLE = mle_gene3
    ), 
  fstat_calc, pheno[, 3]
  ) %>% 
  flatten_dbl() %>% 
  pf(2, 344 - 3, lower.tail = FALSE)
# gene 4
mle_gene4 <- map2(data.frame(x_a), data.frame(x_d), MLE, as.numeric(as.matrix(pheno[, 4])))
pval_gene4 <- pmap(
  list(
    xa = data.frame(x_a), 
    xd = data.frame(x_d), 
    MLE = mle_gene4
    ), 
  fstat_calc, pheno[, 4]
  ) %>% 
  flatten_dbl() %>% 
  pf(2, 344 - 3, lower.tail = FALSE)
# gene 5
mle_gene5 <- map2(data.frame(x_a), data.frame(x_d), MLE, as.numeric(as.matrix(pheno[, 5])))
pval_gene5 <- pmap(
  list(
    xa = data.frame(x_a), 
    xd = data.frame(x_d), 
    MLE = mle_gene5
    ), 
  fstat_calc, pheno[, 5]
  ) %>% 
  flatten_dbl() %>% 
  pf(2, 344 - 3, lower.tail = FALSE)


hist(pval_gene1, breaks = 20, 
     xlab = "p-values", 
     ylab = "Number of p-values", 
     main = "Distribution of p-values")

pval_full_data <- as.data.frame(
  cbind(
    x = 1:ncol(geno), 
    pval_gene1,
    pval_gene2,
    pval_gene3,
    pval_gene4,
    pval_gene5
    )
  )
row.names(pval_full_data) <- colnames(geno)
pval_full_data %>% 
  gather("gene", "pval", 2:6) %>% 
  ggplot(aes(x = pval, fill = gene)) +
  geom_histogram(position = "identity", alpha = 0.5, bins = 50)

pval_full_data %>% 
  gather("gene", "pval", 2:6) %>% 
  ggplot(aes(x = x, y = -log(pval, base = 10), color = gene)) +
  geom_point(alpha = 0.5, size = 2) +
  ggtitle("Manhattan Plot, All Genes") +
  xlab("Genotype Position") +
  ylab("-log10(pVal)")
pval_full_data %>% 
  rownames_to_column("snp") %>% 
  dplyr::select(-x) %>% 
  gather("gene", "pval", 2:6) %>% 
  filter(pval < (0.5 / ncol(geno)))

data_for_qqplot <- data.frame(
  cbind(
    expected = sort(-log10(seq(from = 0,to = 1, length.out = length(pval_full_data$pval_gene1)))),
    gene1_obs_sort = sort(-log10(pval_full_data$pval_gene1)),
    gene2_obs_sort = sort(-log10(pval_full_data$pval_gene2)),
    gene3_obs_sort = sort(-log10(pval_full_data$pval_gene3)),
    gene4_obs_sort = sort(-log10(pval_full_data$pval_gene4)),
    gene5_obs_sort = sort(-log10(pval_full_data$pval_gene5))
    )
  )
data_for_qqplot %>% 
  gather("gene", "neglog10_pval", 2:6) %>% 
  ggplot(aes(x = expected, y = neglog10_pval, color = gene)) +
  geom_point(alpha = 0.5) +
  geom_abline(intercept = 0, slope = 1, col = "black", alpha = 0.5, size = 2)





#lm_test <- lm(pheno$ENSG00000164308.12 ~ x_a[, 1000] + x_d[, 1000] + factor(covars$Population))
#lm_tidy <- tidy(lm_test)
#fstat <- summary(lm_test)$fstatistic
#fstat_2 <- glance(lm_test)$statistic
#pf(fstat[1], fstat[2], fstat[3], lower.tail = FALSE)
```

Seems like there is a problem with just the base model. Based of the PCA analysis, it appears that the Population may be an important covariate to include in the models.

But first, let's test the relationship of each phenotype with each covariate:

```{r}
gene1_pop_lm <- lm(pheno[, 1] ~ as.numeric(factor(covars$Population)))  # significant
gene2_pop_lm <- lm(pheno[, 2] ~ as.numeric(factor(covars$Population)))  # ns
gene3_pop_lm <- lm(pheno[, 3] ~ as.numeric(factor(covars$Population)))  # ns
gene4_pop_lm <- lm(pheno[, 4] ~ as.numeric(factor(covars$Population)))  # ns
gene5_pop_lm <- lm(pheno[, 5] ~ as.numeric(factor(covars$Population)))  # ns


nested_pheno <- pheno %>% 
  rownames_to_column("samples") %>% 
  left_join(
    covars %>% 
      rownames_to_column("samples"),
    by = "samples"
  ) %>% 
  gather("gene", "abundance", 2:6) %>% 
  nest(-gene)

num_factor_pop_lm <- nested_pheno %>% 
  mutate(model = map(data, ~lm(abundance ~ as.numeric(factor(Population)), data = .))) %>% 
  mutate(glance_model = map(model, glance)) %>% 
  unnest(glance_model) %>% 
  dplyr::select(gene, p.value)
num_factor_pop_lm[which(num_factor_pop_lm$p.value < 0.05), ]
# what if Population was treated as a factor and R was allowed to code it
cat_factor_pop_lm <- nested_pheno %>% 
  mutate(model = map(data, ~lm(abundance ~ factor(Population), data = .))) %>% 
  mutate(glance_model = map(model, glance)) %>% 
  unnest(glance_model) %>% 
  dplyr::select(gene, p.value)
cat_factor_pop_lm[which(cat_factor_pop_lm$p.value < 0.05), ]
# what about the Sex covariable?
covar_sex_lm <- nested_pheno %>% 
  mutate(model = map(data, ~lm(abundance ~ factor(Sex), data = .))) %>% 
  mutate(glance_model = map(model, glance)) %>% 
  unnest(glance_model) %>% 
  dplyr::select(gene, p.value)
any(covar_sex_lm < 0.05)
# none are significant on the sex covariable
# what about a combination? 
covars_sex_num_pop_lm <- nested_pheno %>% 
  mutate(model = map(data, ~lm(abundance ~ as.numeric(factor(Population)) + factor(Sex), data = .))) %>% 
  mutate(glance_model = map(model, glance)) %>% 
  unnest(glance_model) %>% 
  dplyr::select(gene, p.value)
covars_sex_num_pop_lm[which(covars_sex_num_pop_lm$p.value < 0.05), ]
nested_pheno %>% 
  mutate(model = map(data, ~lm(abundance ~ as.numeric(factor(Population)) + factor(Sex), data = .))) %>% 
  mutate(tidy_model = map(model, tidy)) %>% 
  unnest(tidy_model) %>% 
  filter(gene %in% covars_sex_num_pop_lm[which(covars_sex_num_pop_lm$p.value < 0.05), 1])
# indicates that the relationship between population only is significant
covars_sex_cat_pop_lm <- nested_pheno %>% 
  mutate(model = map(data, ~lm(abundance ~ factor(Population) + factor(Sex), data = .))) %>% 
  mutate(glance_model = map(model, glance)) %>% 
  unnest(glance_model) %>% 
  dplyr::select(gene, p.value)
covars_sex_cat_pop_lm[which(covars_sex_cat_pop_lm$p.value < 0.05), ]
```

Including covariates:

```{r}
library(lmtest)

lr_likelihood <- function(y, x_input = NULL){
    n_samples <- length(y)
    X_mx <- cbind(matrix(1, nrow = n_samples, ncol = 1), x_input)
    MLE_beta <- ginv(t(X_mx) %*% X_mx) %*% t(X_mx) %*% y
    y_hat <- X_mx %*% MLE_beta
    var_hat <- sum((y - (y_hat))^2) / (n_samples - 1)
    log_likelihood <- -((n_samples / 2) * log(2 * pi * var_hat) ) - ((1/ (2*var_hat)) * sum((y - (y_hat))^2))
    return(log_likelihood)
}

LRT_test <- function(logl_H0, logl_HA, df_test){

    LRT<-2*logl_HA-2*logl_H0 #likelihood ratio test statistic
    #likelihood ratio test statistic for every genotype
    pval <- pchisq(LRT, df_test, lower.tail = F)
    return(pval)
}

set.seed(2018)
x = sample(c(-1,0,1), 100, replace = TRUE)
y = 0.9 * x + rnorm(100)
h0_nocovar <- lr_likelihood(y)
h1_nocovar <- lr_likelihood(y, x)
LRT_test(h0_nocovar, h1_nocovar, df_test = 1)
x_c = sample(c(0,1), 100, replace = TRUE)
y2 = y + 0.8 * x_c 
  
h0_withcovar <- lr_likelihood(y2)
h1_withcovar <- lr_likelihood(y2, x)
LRT_test(h0_withcovar, h1_withcovar, df_test = 1)

h0_includecovar <- lr_likelihood(y2, x_c)
ha_includecovar <- lr_likelihood(y2, cbind(x,x_c))
LRT_test(h0_includecovar, ha_includecovar, df_test = 1)

# y2 = y, x_c = covar, x = x

X_mat_null <- cbind(1, x_c)
beta_hat_null <- ginv(t(X_mat_null) %*% X_mat_null) %*% t(X_mat_null) %*% y2
X_mat_alt <- cbind(1, x, x_c)
beta_hat_alt <- ginv(t(X_mat_alt) %*% X_mat_alt) %*% t(X_mat_alt) %*% y2
y_hat_null <- X_mat_null %*% beta_hat_null
y_hat_alt <- X_mat_alt %*% beta_hat_alt
SSE_null <- sum((y2 - y_hat_null) ^ 2)
SSE_alt <- sum((y2 - y_hat_alt) ^ 2)
fstat <- ((SSE_null - SSE_alt) / 2) / (SSE_alt / (length(x) - 3))

pf(fstat, 2, length(x) - 3, lower.tail = FALSE)

# testing with real data

X_mat_null <- cbind(matrix(1, nrow = nrow(x_a), ncol = 1), NULL)
beta_hat_null <- ginv(t(X_mat_null) %*% X_mat_null) %*% t(X_mat_null) %*% pheno[, 2]
X_mat_alt <- cbind(1, x_a[, 2], x_d[, 2])
beta_hat_alt <- ginv(t(X_mat_alt) %*% X_mat_alt) %*% t(X_mat_alt) %*% pheno[, 2]
y_hat_null <- X_mat_null %*% beta_hat_null
y_hat_alt <- X_mat_alt %*% beta_hat_alt
SSE_null <- sum((pheno[, 2] - y_hat_null) ^ 2)
SSE_alt <- sum((pheno[, 2] - y_hat_alt) ^ 2)
fstat <- ((SSE_null - SSE_alt) / 2) / (SSE_alt / (nrow(geno) - 3))
pf(fstat, 2, nrow(geno) - 3, lower.tail = FALSE)











```

